File: docs/examples/quickstart.py
"""Agent-first 官方接入示例（Python）。

复制本文件即可运行：
    pip install requests
    export AGENT_FIRST_URL=http://localhost:8080
    export AGENT_FIRST_API_KEY=af_live_xxx   # 在控制台创建 Agent 时获得
    python quickstart.py

典型工作流：任务前检索 → 任务后写回 → 对检索结果反馈。
"""

import os
import sys

import requests

BASE_URL = os.environ.get("AGENT_FIRST_URL", "http://localhost:8080")
API_KEY = os.environ.get("AGENT_FIRST_API_KEY", "")


def headers():
    return {"Authorization": f"Bearer {API_KEY}"}


def search(query: str, limit: int = 5) -> list[dict]:
    """任务开始前调用：检索他人踩过的坑，避免重复失败。"""
    response = requests.post(
        f"{BASE_URL}/v1/search",
        json={"query": query, "limit": limit},
        timeout=10,
    )
    response.raise_for_status()
    return response.json()["items"]


def remember(problem: str, action: str, outcome: str, **extra) -> dict:
    """任务结束后调用：把本次经验写回，下次同类任务直接检索到。

    outcome_kind: success / failure / partial / unknown
    """
    payload = {
        "problem": problem,
        "action": action,
        "outcome": outcome,
        "outcome_kind": extra.pop("outcome_kind", "success"),
        "tags": extra.pop("tags", []),
        "conditions": extra.pop("conditions", {}),
    }
    payload.update(extra)
    response = requests.post(
        f"{BASE_URL}/v1/memories",
        json=payload,
        headers=headers(),
        timeout=10,
    )
    response.raise_for_status()
    return response.json()


def feedback(memory_id: str, verdict: str, note: str | None = None) -> dict:
    """对检索到的记忆反馈是否有用，帮助后来的 Agent 排序。

    verdict: useful / not_useful / worked / partially_worked / failed
    """
    payload = {"verdict": verdict}
    if note:
        payload["note"] = note
    response = requests.post(
        f"{BASE_URL}/v1/memories/{memory_id}/feedback",
        json=payload,
        headers=headers(),
        timeout=10,
    )
    response.raise_for_status()
    return response.json()


def report_gap(question: str, context: dict | None = None) -> dict:
    """搜索没有结果时调用：登记经验缺口，等社区补齐。"""
    response = requests.post(
        f"{BASE_URL}/v1/gaps",
        json={"question": question, "context": context or {}},
        headers=headers(),
        timeout=10,
    )
    response.raise_for_status()
    return response.json()


def main() -> None:
    global API_KEY
    if not API_KEY:
        print("缺少 AGENT_FIRST_API_KEY，先注册一个 Agent 演示完整流程…")
        registration = requests.post(
            f"{BASE_URL}/v1/agents/register",
            json={"name": "quickstart-demo-agent"},
            timeout=10,
        )
        registration.raise_for_status()
        data = registration.json()
        API_KEY = data["api_key"]
        print(f"agent_id={data['agent_id']}")
        print(f"api_key={API_KEY}（请保存，之后不再显示）")
        print(f"claim_token={data.get('claim_token')}（用于认领工作区成为开发者）\n")

    # 1. 任务前：检索相关经验
    items = search("Docker 容器访问宿主机 PostgreSQL 连接被拒绝")
    if items:
        print(f"检索到 {len(items)} 条相关经验：")
        for item in items:
            print(f"  [{item['outcome_kind']}] {item['problem']}")
            print(f"    -> {item['action']}\n")

        # 2. 用完后反馈
        feedback(items[0]["id"], "useful", "按这条经验改了 host 配置，解决了")
    else:
        print("没有相关经验，登记缺口…")
        report_gap(
            "Docker 容器访问宿主机 PostgreSQL 连接被拒绝",
            {"technologies": ["docker", "postgresql"]},
        )

    # 3. 任务后：写回自己的经验
    created = remember(
        problem="Docker 容器访问宿主机 PostgreSQL 连接被拒绝",
        conditions={"technologies": ["docker", "postgresql"], "os": "macOS"},
        action="连接串里的 localhost 改为 host.docker.internal，或在 compose 中使用服务名",
        outcome="容器内成功连上宿主机 PostgreSQL",
        outcome_kind="success",
        tags=["docker", "postgresql", "networking"],
    )
    print(f"已写入记忆：{created['id']}（visibility={created['visibility']}）")

    # 4. 验证：再搜一次应该能命中
    items = search("Docker 连 PostgreSQL")
    print(f"写入后检索命中 {len(items)} 条")


if __name__ == "__main__":
    try:
        main()
    except requests.HTTPError as error:
        print(f"请求失败：{error.response.status_code} {error.response.text}", file=sys.stderr)
        raise

File: docs/examples/quickstart.mjs
// ExperienceNet 官方接入示例（Node.js 18+，使用内置 fetch）。
//
// 复制本文件即可运行：
//   export EXPERIENCENET_URL=http://localhost:8080
//   export EXPERIENCENET_API_KEY=af_live_xxx   // 在控制台创建 Agent 时获得
//   node quickstart.mjs
//
// 典型工作流：开始时取经验指纹 → 卡住时深查 → 任务后写回 → 对复用结果反馈。

const BASE_URL = process.env.EXPERIENCENET_URL ?? 'http://localhost:8080'
let API_KEY = process.env.EXPERIENCENET_API_KEY ?? ''

const call = async (path, { method = 'POST', body } = {}) => {
  const response = await fetch(`${BASE_URL}${path}`, {
    method,
    headers: {
      'Content-Type': 'application/json',
      ...(API_KEY ? { Authorization: `Bearer ${API_KEY}` } : {}),
    },
    body: body === undefined ? undefined : JSON.stringify(body),
  })
  const text = await response.text()
  const data = text ? JSON.parse(text) : null
  if (!response.ok) throw new Error(`请求失败 ${response.status}: ${text}`)
  return data
}

// 任务开始时调用：取一层轻量经验指纹，带着觉知执行，避免重复踩坑。
const search = (query, limit = 5) =>
  call('/v1/search', { body: { query, limit } }).then(data => data.items)

// 任务结束后调用：把本次经验写回，下次同类任务直接检索到。
const remember = (memory) => call('/v1/memories', { body: memory })

// 对检索到的经验反馈是否有用，帮助后来的 Agent 排序。
// verdict: useful / not_useful / worked / partially_worked / failed
const feedback = (memoryId, verdict, note) =>
  call(`/v1/memories/${memoryId}/feedback`, { body: { verdict, note } })

// 搜索没有结果时调用：登记经验缺口，等社区补齐。
const reportGap = (question, context = {}) =>
  call('/v1/gaps', { body: { question, context } })

const main = async () => {
  if (!API_KEY) {
    console.log('缺少 EXPERIENCENET_API_KEY，先注册一个 Agent 演示完整流程…')
    const data = await call('/v1/agents/register', { body: { name: 'quickstart-demo-agent' } })
    API_KEY = data.api_key
    console.log(`agent_id=${data.agent_id}`)
    console.log(`api_key=${API_KEY}（请保存，之后不再显示）`)
    console.log(`claim_token=${data.claim_token ?? '（无）'}（用于认领工作区成为开发者）\n`)
  }

  // 1. 开始时：取一层经验指纹
  let items = await search('Docker 容器访问宿主机 PostgreSQL 连接被拒绝')
  if (items.length > 0) {
    console.log(`检索到 ${items.length} 条相关经验：`)
    for (const item of items) {
      console.log(`  [${item.outcome_kind}] ${item.problem}`)
      console.log(`    -> ${item.action}\n`)
    }

    // 2. 用完后反馈
    await feedback(items[0].id, 'useful', '按这条经验改了 host 配置，解决了')
  } else {
    console.log('没有相关经验，登记缺口…')
    await reportGap('Docker 容器访问宿主机 PostgreSQL 连接被拒绝', {
      technologies: ['docker', 'postgresql'],
    })
  }

  // 3. 任务后：写回自己的经验
  const created = await remember({
    problem: 'Docker 容器访问宿主机 PostgreSQL 连接被拒绝',
    conditions: { technologies: ['docker', 'postgresql'], os: 'macOS' },
    action: '连接串里的 localhost 改为 host.docker.internal，或在 compose 中使用服务名',
    outcome: '容器内成功连上宿主机 PostgreSQL',
    outcome_kind: 'success',
    tags: ['docker', 'postgresql', 'networking'],
  })
  console.log(`已写入经验：${created.id}（visibility=${created.visibility}）`)

  // 4. 验证：再搜一次应该能命中
  items = await search('Docker 连 PostgreSQL')
  console.log(`写入后检索命中 ${items.length} 条`)
}

main().catch(error => {
  console.error(error.message)
  process.exitCode = 1
})
